### PR TITLE
SpringBoot API for Frontend-Database Communication

### PR DESCRIPTION
**Pull request by vibhu18116**
The API serves as a medium of communication between frontend and
Database for fetching and updating cart and order details.

### CODE REVIEW

File: backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class ItemDetails{
+	public String userID;
+	public String itemName;
+	public int quantity = 0;
+
+	@Override
+	public String toString(){
+		return userID;
+	}
+}
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java:18]: Move data class as separate class and move all the **public** fields to **private** and provide the getters/setters for the same.

File: backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+@RestController
+public class CartDocumentsController {
+
+	@RequestMapping(value="/getItemsInCart")
+	public ArrayList<CartItem> getItemsInCart(@RequestParam String user)
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java:24]: Code always against interface. Use `List` not `ArrayList`.

File: backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			throws InterruptedException, ExecutionException {
+		ArrayList<CartItem> allItemsInCart = SMBKiranaApplication.fireStoreInstance.retrieveCartItems(user);
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java:26]: Code always against interface. Use List not ArrayList.

File: backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		return allItemsInCart;
+	}
+
+	@RequestMapping(value="/updateItemQuantity", method=RequestMethod.POST)
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java:30]: Prefer using static import for `POST`

File: backend/src/main/java/com/googleinterns/smb/CartItem.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.googleinterns.smb;
+
+public class CartItem{
+    private String item_name;
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/CartItem.java:4]: Use Camel Case variable names

File: backend/src/main/java/com/googleinterns/smb/FirebaseInitialiser.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+        String docID = addedDocRef.get().getId();
+        List<Item> allItems = orderDoc.getAllItems();
+
+        for (Item item: allItems){
+            ApiFuture<DocumentReference> addedSubDocRef = db.collection("order").document(docID).collection("items").add(item);  
+        }
+    }
+
+}
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/FirebaseInitialiser.java:143]: Run Linter on this class

File: backend/src/main/java/com/googleinterns/smb/Item.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package com.googleinterns.smb;
+
+public class Item {
+    
+    public String item_name;
+    public int quantity;
+    public float unit_price;
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/Item.java:7]: Camel case variable names please!

File: backend/src/main/java/com/googleinterns/smb/Order.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        this.order_status = "PROCESSING";
+    }
+
+    public Order(String userID, String merchantName, String merchantAddress, String offers){
+        this.user_id = userID;
+        this.order_id = generateOrderID();
+        this.merchant_name = merchantName;
+        this.merchant_address = merchantAddress;
+        this.offers_availed = offers;
+        this.order_status = "PROCESSING";
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/Order.java:36]: Move it to ENUM

File: backend/src/main/java/com/googleinterns/smb/Order.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            StringBuilder sb = new StringBuilder(5);
+
+            for (int i = 0; i < 5; i++) sb.append(randomDigits[i]);
+            String randomID = epochTime + ":" + sb.toString();
+            return randomID;
+ 
+		} catch (ParseException e) {
+			return "ERROR";
+		}
+    }
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/Order.java:68]: Seems super complex, please simplify

File: backend/src/main/java/com/googleinterns/smb/OrderDocumentsController.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    public String userID;
+}
+
+class OrderGeneration{
+    public String userId;
+    public String servicingMerchantName;
+    public String servicingMerchantAddress;
+    public String offers;
+    public List<Item> itemDetails;
+}
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/OrderDocumentsController.java:18]: Move to separate data class

File: backend/src/main/java/com/googleinterns/smb/SMBKiranaApplication.java [hunk 11]
```diff
@@ -1,26 +1,21 @@
...
 
-import java.util.HashMap;
-import java.util.Map;
 
 @SpringBootApplication
-@RestController
 public class SMBKiranaApplication {
-	
+
+	public static FirebaseInitialiser fireStoreInstance;
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/SMBKiranaApplication.java:11]: Wrong way! Use Dependency injection

File: backend/src/main/java/com/googleinterns/smb/SpringBootSMB.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        // *** URL below needs to match the Vue client URL and port ***
+        config.setAllowedOrigins(Collections.singletonList("*")); 
+        config.setAllowedMethods(Collections.singletonList("*"));  
+        config.setAllowedHeaders(Collections.singletonList("*"));  
+        source.registerCorsConfiguration("/**", config);  
+        FilterRegistrationBean bean = new FilterRegistrationBean<>(new CorsFilter(source));
+        bean.setOrder(Ordered.HIGHEST_PRECEDENCE);  
+        return bean;  
+      } 
+}
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/SpringBootSMB.java:29]: Use @CrossOrigin

File: backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java (hunk 1 above)
Comment (vibhu18116) [reply backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java]: Done

File: backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import org.springframework.web.bind.annotation.*;
+
+class ItemDetails{
+	public String userID;
+	public String itemName;
+	public int quantity = 0;
```
Comment (pulkit-singhal) [inline backend/src/main/java/com/googleinterns/smb/CartDocumentsController.java]: *int* is default to zero, so no initialisation required.
As a good practice, all the field should be **private** and can have **public** getter setters

**Pull request merged**